src/Halogen/Query/EventSource.purs

```diff
@@ -1,187 +1,150 @@
-  ( EventSource(..)
-  , SubscribeStatus(..)
-  , unEventSource
-  , interpret
-  , hoist
-  , eventSource
-  , eventSource'
-  , eventSource_
-  , eventSource_'
-  , catMaybes
-  , produce
-  , produce'
-  , produceAff
-  , produceAff'
-  ) where
+module Halogen.Query.EventSource where
 
 import Prelude
 
 import Control.Coroutine as CR
-import Control.Monad.Aff (Aff, attempt, forkAff, runAff_)
+import Control.Monad.Aff (Aff, attempt, launchAff_)
 import Control.Monad.Aff.AVar (AVAR)
 import Control.Monad.Aff.AVar as AV
 import Control.Monad.Aff.Class (class MonadAff, liftAff)
 import Control.Monad.Eff (Eff)
 import Control.Monad.Eff.Class (liftEff)
 import Control.Monad.Eff.Exception as Exn
 import Control.Monad.Free.Trans as FT
-import Control.Monad.Rec.Class as Rec
 import Control.Monad.Trans.Class (lift)
 import Data.Bifunctor (lmap)
-import Data.Either (Either(..), either)
-import Data.Foldable (for_)
-import Data.Maybe (Maybe)
+import Data.Either (Either(..))
+import Data.Monoid (class Monoid)
+import Data.Profunctor (dimap)
 
-newtype EventSource f m = EventSource (m { producer :: CR.Producer (f SubscribeStatus) m Unit, done :: m Unit })
+-- | An event source definition - an effect in `m` that when run returns a
+-- | producer coroutine that emits queries of type `f`, and runs in the effect
+-- | monad `m`.
+-- |
+-- | It's generally unnecessary to build values of this type directly with this
+-- | constructor, the `affEventSource` and `effEventSource` cover the most
+-- | event source constructions.
+newtype EventSource m f = EventSource (m { producer :: CR.Producer (f Unit) m Unit, finalizer :: Finalizer m })
 
-unEventSource :: forall f m. EventSource f m -> m { producer :: CR.Producer (f SubscribeStatus) m Unit, done :: m Unit }
-unEventSource (EventSource e) = e
+-- | Constructs an event source from a setup function that operates in `Aff`.
+-- |
+-- | - The `Emitter` that the passed function receives is used to `emit` queries
+-- |   that will be received by the current component, or can be `close`d to
+-- |   shut down the event source and remove the subscription.
+-- | - The `Finalizer` that the passed function produces is there to allow for
+-- |   some clean-up action to be taken when the event source is unsubscribed
+-- |   from. This also runs if the `Emitter` is `close`d. `mempty` can be used
+-- |   here if there is no clean-up to perform.
+affEventSource
+  :: forall m f eff
+   . MonadAff (avar :: AVAR | eff) m
+  => (Emitter (Aff (avar :: AVAR | eff)) f -> Aff (avar :: AVAR | eff) (Finalizer (Aff (avar :: AVAR | eff))))
+  -> EventSource m f
+affEventSource recv = EventSource $ liftAff do
+  inputVar <- AV.makeEmptyVar
+  finalizeVar <- AV.makeEmptyVar
+  let
+    producer = do
+      lift $ liftAff $ flip AV.putVar finalizeVar =<< recv (Emitter (flip AV.putVar inputVar))
+      CR.producer $ liftAff $ AV.takeVar inputVar
+    finalizer = Finalizer do
+      liftAff (attempt (AV.takeVar finalizeVar)) >>= case _ of
```

Comment:
Why attempt is needed?

Reply:
In case the finalizer is run a second time - it kills the var, so attempting to `take` from it would result in an `Aff` error.